### DISCUSSION TITLE
Can we use Technitium DNS Server for private host and domains

### DISCUSSION THREAD

**Discussion by ciscohack** [TechnitiumSoftware/DnsServer#846] [General]
Hi Team,

Can we use Technitium DNS Server for private hosts and domains? If yes can someone guide me steps how to do that

**Comment by ShreyasZare** [comment]
Thanks for asking. If you mean self hosting domain names then take a look at [this blog post](https://blog.technitium.com/2022/06/how-to-self-host-your-own-domain-name.html). If you mean to host private local domain then you can do that by creating a zone for your private domain name from the admin panel's Zone section.

**Comment by ciscohack** [comment]
@ShreyasZare Thank you so much. Yes, I using the technetium DNS Docker version for private DNS where I want to resolve my local domain and network device names to IP address for this. I created a zone and added A record with PTR record created but DNS is not working and not resolving and failing. I followed the documentation but this tool is amazing but documentation lacks... I need your help to make this work. I get the server failure RCODE 

<img width="1137" alt="image" src="https://github.com/TechnitiumSoftware/DnsServer/assets/11479557/93847928-0684-4c1c-a6a5-d6e135ddba2c">

**Reply by ShreyasZare** [reply]
Thanks for the details of the zone. Also provide output of the DNS Client tab for the query you are trying to resolve.

**Reply by ciscohack** [reply]
Thanks for the quick response, my friend.. So my server name is web-SRV with IP 20.1.1.5 which matched in A record... But I tried another domain name but all failing for me..

<img width="1003" alt="image" src="https://github.com/TechnitiumSoftware/DnsServer/assets/11479557/c3913080-5da6-40a0-9593-b5bac3af5036">

<img width="711" alt="image" src="https://github.com/TechnitiumSoftware/DnsServer/assets/11479557/1fbf4417-2930-49b9-a343-b1794c09898b">

**Reply by ciscohack** [reply]
```
{
  "Metadata": {
    "NameServer": "20.1.1.2",
    "Protocol": "Udp",
    "DatagramSize": "101 bytes",
    "RoundTripTime": "1803.53 ms"
  },
  "EDNS": {
    "UdpPayloadSize": 1232,
    "ExtendedRCODE": "ServerFailure",
    "Version": 0,
    "Flags": "None",
    "Options": [
      {
        "Code": "EXTENDED_DNS_ERROR",
        "Length": "49 bytes",
        "Data": {
          "InfoCode": "NoReachableAuthority",
          "ExtraText": "No response from name servers for web-srv. A IN"
        }
      },
      {
        "Code": "EXTENDED_DNS_ERROR",
        "Length": "2 bytes",
        "Data": {
          "InfoCode": "CachedError",
          "ExtraText": null
        }
      },
      {
        "Code": "EXTENDED_DNS_ERROR",
        "Length": "2 bytes",
        "Data": {
          "InfoCode": "StaleAnswer",
          "ExtraText": null
        }
      }
    ]
  },
  "DnsClientExtendedErrors": [
    {
      "InfoCode": "NetworkError",
      "ExtraText": "20.1.1.2 returned RCODE=ServerFailure for web-srv. A IN"
    }
  ],
  "Identifier": 27089,
  "IsResponse": true,
  "OPCODE": "StandardQuery",
  "AuthoritativeAnswer": false,
  "Truncation": false,
  "RecursionDesired": true,
  "RecursionAvailable": true,
  "Z": 0,
  "AuthenticData": false,
  "CheckingDisabled": false,
  "RCODE": "ServerFailure",
  "QDCOUNT": 1,
  "ANCOUNT": 0,
  "NSCOUNT": 0,
  "ARCOUNT": 1,
  "Question": [
    {
      "Name": "web-SRV",
      "Type": "A",
      "Class": "IN"
    }
  ],
  "Answer": [],
  "Authority": [],
  "Additional": [
    {
      "Name": "",
      "Type": "OPT",
      "Class": "1232",
      "TTL": "0 (0 sec)",
      "RDLENGTH": "65 bytes",
      "RDATA": {
        "Options": [
          {
            "Code": "EXTENDED_DNS_ERROR",
            "Length": "49 bytes",
            "Data": {
              "InfoCode": "NoReachableAuthority",
              "ExtraText": "No response from name servers for web-srv. A IN"
            }
          },
          {
            "Code": "EXTENDED_DNS_ERROR",
            "Length": "2 bytes",
            "Data": {
              "InfoCode": "CachedError",
              "ExtraText": null
            }
          },
          {
            "Code": "EXTENDED_DNS_ERROR",
            "Length": "2 bytes",
            "Data": {
              "InfoCode": "StaleAnswer",
              "ExtraText": null
            }
          }
        ]
      },
      "DnssecStatus": "Disabled"
    }
  ]
}

**Reply by ShreyasZare** [reply]
You need to use `web-srv.mydnszone` domain name to query. The `web-srv` is subdomain for that zone so you must use full domain name.  

Also, you are not querying the local DNS server but using `System DNS` which is whatever DNS server is configured on your LAN config.

The error "Cannot assign requested address" when you select `This Server` for query is also an issue. It seems to be some config issue you have. You should check the DNS Logs from the web panel and find the complete error log in there. Post any error log you see here.

**Reply by ciscohack** [reply]
@ShreyasZare that's what the confusion what are the local DNS server and system-DNS server differences? it's unclear to me and I didn't find documentation explaining in this tool context.  I have in my host device I have configured the Technitium server IP as the DNS server IP. I want Technitium DNS to resolve all my DNS queries...now it falls into the local DNS server or system-DNS unsure to me. 

second, suppose I have a website named example.com and if I create an A record in my zone name mydnszone so now to resolve the domain name to IP  in the address bar should I put a URL like this http://www.example.com.mydnszone to make it work?

As per your suggestion when I tried web-srv.mydnszone it works but still with some communication errors but there is no error log in the DNS server

<img width="684" alt="image" src="https://github.com/TechnitiumSoftware/DnsServer/assets/11479557/3cbd1a63-4879-4613-933e-38afadcd5a62">

**Reply by ShreyasZare** [reply]
Let me just clear the confusion here. The DNS Client tab on the admin panel is an independent tool for testing. The "This Server" entry in there means to query the locally running DNS server. And "System DNS" entry means to use the DNS server configured on your network adapter. You can have a different DNS server configured on the network adapter since its not necessary to have local DNS to be used on the same server.

If you want to host "example.com" then you have to create a zone for "example.com". Adding "www.example.com" entry in "mydnszone" zone is not going to work since it another domain name and not "example.com".

For the nslookup output, the "communication error" in there is just a failure to reverse lookup the server's IP address "20.1.1.2" and not related to lookup your "web-srv.mydnszone" domain name. The nslookup tool first does reverse lookup for the DNS server's address to show you its domain name in the output. If that fails, it just shows the server's IP. After that it will query for your domain name. So, its not really an issue with resolving your domain name.

Also, if you are using "20.1.1.1" IP in your private network then you should not do that. Use private IP address range instead. If you use public IP range in your private network then you wont be able to visit a website that is using public IP address in the same range since your computer thinks that the website is located in your private LAN.

**Reply by ciscohack** [reply]
@ShreyasZare Thanks a lot for the detailed explanation. if I understand correctly  "this server/system DNS server" suppose if I used Technitium DNS Server as my DNS query resolver for my client machines then this server and system server is same and I should resolve the query right. so if my understanding is correct then in screenshot error for this server is for same condition only

**Reply by ShreyasZare** [reply]
The "This Server" and "System DNS" only applies to the DNS Client tool that is running on your server. I think the "System DNS" thing is confusing you the most. 

Lets say you have a server on which you have installed the DNS server. On that server, you can configure the OS to use `127.0.0.1` as the DNS server so that the OS uses the same DNS server you have installed. But imagine that you configure your OS to use `9.9.9.9` public resolver instead. So, in the DNS Client tab, when you select "This Server", the tool sends the requests to `127.0.0.1` and when you select "System DNS" it will send the request to `9.9.9.9` which your OS is configured to use.

For clients on your network, they will be configured to use the IP address of your DNS server. The clients do not have anything to do with the concept of "This Server" or "System DNS". Those concepts apply only for the DNS Client tool.

**Reply by ciscohack** [reply]
thanks doubt cleared on this this.. I feel this amazing tool documentation can be much more better

**Reply by ShreyasZare** [reply]
You're welcome. Yes, there is requirement for documentation but due to time constrains its not getting prioritized.

**Reply by ciscohack** [reply]
Yes I can understand..Thanks once again

**Reply by ciscohack** [reply]
@ShreyasZare I think we need to fix this appstore error also.. it's not working anywhere docker/linux or other installation place

<img width="792" alt="image" src="https://github.com/TechnitiumSoftware/DnsServer/assets/11479557/3d0de289-1a90-4caf-8372-ca22fa5bb1f0">

**Reply by ShreyasZare** [reply]
Its some local issue causing this since the service is running well as expected. Check the DNS logs once from the admin panel to find error log for this. The error log will explain why the domain failed to resolve.

**Reply by ciscohack** [reply]
Ok thanks

**Reply by ciscohack** [reply]
@ShreyasZare One question suppose I have created a zone on Technitium DNS and A records for my internal host but now I want my Technitium DNS server able to resolve public domain as well like google.com/facbook.com etc. So for this case, we can configure forward address right to some external DNS server for recursive lookup. How I achieve this now my client machines having Technitium DNS server IP able to resolve internal host but failed for public domain.. there is conditional forwarding option but not able to understand it fully.

**Reply by ShreyasZare** [reply]
The DNS server will automatically resolve other domain names recursively unless you have disabled recursion from the Settings > Recursion section.

Configuring forwarders is not necessary as the DNS server can resolve recursively. But, recursion may not work with all ISPs due to blocking and forwarding using encrypted DNS protocols may be required.

**Reply by ciscohack** [reply]
@ShreyasZare oh is it.. then something wrong it's not working for me

**Reply by ShreyasZare** [reply]
Test querying any domain name using the DNS Client tab on the admin panel and post the output here. That will give clue on what the issue is.

**Reply by ciscohack** [reply]
@ShreyasZare ok ..will do it

**Comment by Hemsby** [comment]
Edited because I read the above incorrectly.

Unless your DNS Search suffix is mydnszone then without the FQDN this will not work by default.  If it is using .mydnszone as an appended suffix then the address bar will still not work.  You would create a CNAME record called example pointing to example.com and the address bar would be http://example.mydnszone .  If you want the WWW then again you would create a CNAME WWW pointing to www.example.com and the address would be http://www.example.mydnszone

**Reply by ciscohack** [reply]
ok so basically to make it work like example.com. I have to create CNAME record to alias the http://example.mydnszone right other then this no other way

**Reply by Hemsby** [reply]
I think you have misunderstood, or your example was incorrect.  I have some time, so if you want you can email me and I can help you, via Teams or Skype etc.

**Reply by ciscohack** [reply]
@Hemsby  Oh, I missed your email. I am available if you have time we can connect.. also what is the solution to fix below error.. teams code if you provide will join.. if not now we can connect tomorrow anytime suitable to you

<img width="796" alt="image" src="https://github.com/TechnitiumSoftware/DnsServer/assets/11479557/0ce5f424-1059-4862-9c9a-882cb48b6307">

**Reply by Hemsby** [reply]
Sorry, also missed yours!  Send me and email and we can arrange some time.

**Reply by ciscohack** [reply]
Thanks @Hemsby. No problem I have sent you an mail. whenever you will have time let me know over mail then we will connect.

Thanks once again

**Comment by ciscohack** [comment]
> I think you have misunderstood, or your example was incorrect. I have some time, so if you want you can email me and I can help you, via Teams or Skype etc.

@Hemsby let me know my friend whenever you have time to connect .. please do send me email response. Thanks to you as well for all help so far
